packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/PropertiesGrouping.ts

```diff
@@ -0,0 +1,332 @@
+import { ECClass, IMetadataProvider } from "../../../Metadata";
+import { createDefaultValueFormatter } from "../../../values/Formatting";
+import { TypedPrimitiveValue } from "../../../values/Values";
+import { getClass } from "../../Common";
+import { GroupingHandler, GroupingHandlerResult } from "../Grouping";
+
+interface PropertyInfo {
+  label: string;
+  propertyName: string;
+  fullClassName: string;
+  extraValues?:
+    | { formattedValue: string }
+    | {
+        fromValue: number;
+        toValue: number;
+      };
+}
+
+interface PropertyGroupingInformation {
+  ungrouped: ProcessedInstanceHierarchyNode[];
+  grouped: Map<string, { property: PropertyInfo; groupedNodes: ProcessedInstanceHierarchyNode[] }>;
+}
+
+/** @internal */
+export interface PropertyGroupInfo {
+  ecClass: ECClass;
+  previousPropertiesGroupingInfo: PreviousPropertiesGroupingInfo;
+  propertyGroup: Omit<PropertyGroup, "propertyValue">;
+}
+
+/** @internal */
+export type PreviousPropertiesGroupingInfo = Array<{ fullClassName: string; propertyGroup: Omit<PropertyGroup, "propertyValue"> }>;
+
+/** @internal */
+export async function createPropertyGroups(
+  metadata: IMetadataProvider,
+  nodes: ProcessedInstanceHierarchyNode[],
+  propertyInfo: PropertyGroupInfo,
+): Promise<GroupingHandlerResult> {
+  const groupings: PropertyGroupingInformation = { ungrouped: [], grouped: new Map() };
+  for (const node of nodes) {
+    const byProperties = node.processingParams?.grouping?.byProperties;
+    if (!byProperties) {
+      groupings.ungrouped.push(node);
+      continue;
+    }
+    if (!(await shouldCreatePropertyGroup(metadata, propertyInfo, byProperties, node.key.instanceKeys[0].className))) {
+      groupings.ungrouped.push(node);
+      continue;
+    }
+    const currentProperty = byProperties.propertyGroups[propertyInfo.previousPropertiesGroupingInfo.length];
+    const propertyToGroup = {
+      propertyName: currentProperty.propertyName,
+      fullClassName: byProperties.fullClassName,
+    };
+
+    if (currentProperty.ranges) {
+      if (typeof currentProperty.propertyValue !== "number") {
+        addGroupingToMap(
+          groupings.grouped,
+          `${currentProperty.propertyName}:Other`,
+          {
+            label: "Other",
+            ...propertyToGroup,
+          },
+          node,
+        );
+        continue;
+      }
+      const propValue = currentProperty.propertyValue;
+      const matchingRange = currentProperty.ranges.find((range) => propValue >= range.fromValue && propValue <= range.toValue);
+
+      if (matchingRange) {
+        const rangeLabel = matchingRange.rangeLabel ?? `${matchingRange.fromValue} - ${matchingRange.toValue}`;
+        addGroupingToMap(
+          groupings.grouped,
+          `${currentProperty.propertyName}:${matchingRange.fromValue} - ${matchingRange.toValue}(${rangeLabel})`,
```

Comment:
there are a whole bunch of different formats for serializing grouping information - could we have a single function that does it?

Reply:
Refactored the code (didn't add a function for this serialization is now quite simple).